doc: describe upgrade and recovery for raft topology

piodul opened the pull request:
Document the manual upgrade procedure that is required to enable
consistent cluster management in clusters that were upgraded from an
older version to ScyllaDB Open Source 6.0. This instruction is placed in
previously placeholder "Enable Raft-based Topology" page which is a part
of the upgrade instructions to ScyllaDB Open Source 6.0.

Add references to the new description in the "Raft Consensus Algorithm
in ScyllaDB" document in relevant places.

Extend the "Handling Node Failures" document so that it mentions steps
required during recovery of a ScyllaDB cluster running version 6.0.

Fixes: https://github.com/scylladb/scylladb/issues/17341

kbr-scylla commented:
You can rebase on `master` now.

@nuivall please review too (only the last commit; the previous commits are from PR that was already merged)
especially in the context of whether it's necessary to extend parts of it with some auth-v2 related stuff (e.g. preconditions such as don't perform auth statements etc.)

kbr-scylla commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+
+	curl -X POST "http://127.0.0.1:10000/storage_service/raft_topology/upgrade"
+
+Next, wait until all nodes report that upgrade is complete. Checking that a single node finished upgrade can be done in a number of ways:
```
Too many options, instructions unclear, ~~p~~hand stuck in toaster

Maybe let's distinguish one of the options as recommended one, and the other options as optional?

piodul replied:
While looking at the logs is probably the least reliable way to do it and I can drop it, I'm not sure about whether CQL or HTTP is better.

piodul replied:
I removed the log-gazing method and left CQL + HTTP as two equally valid options. Let me know if you would like me to promote one over the other.

kbr-scylla commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+What if upgrade gets stuck?
+===========================
+
+If the process gets stuck at some point, perform :ref:`the Raft recovery procedure <recovery-procedure>`.
```
Before referring to the recovery procedure, we should ask the admin to perform the standard checks, i.e. are all nodes UP? restart them if you can etc.

kbr-scylla commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+
+      raft_topology - upgrade to raft topology has finished
+
+After the upgrade is complete on all nodes, wait for the duration of ``ring_delay_ms`` before issuing any topology changes in order to avoid data loss from writes that were started before the upgrade.
```
No point referring to `ring_delay_ms`. Let's just say "one minute"?

nuivall commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+
+  Alternatively, this can be checked programmatically by checking whether ``value`` column under the key ``enabled_features`` contains the name of the feature in the ``system.scylla_local`` table. This can be done with the following query:
+
+  .. code-block:: none
```
perhaps additionally we could add full bash example? It'd definitely save admins some effort

piodul replied:
Hmm, perhaps I can include a snippet like `cqlsh -e "...query string..."`, like we do with `curl` for the HTTP endpoint.

piodul replied:
Ah, right, this is not only about querying but also checking whether the feature name is there. It makes even more sense to provide a bash command

nuivall replied:
I guess the first user of this will be cloud which'd copy it into siren codebase

zimnx commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+.. _enable-raft-topology-6.0-prerequisites:
+
+Prerequisites
+=============
```
@tnozicka seems there's an way to upgrade existing clusters to use consistent topology changes, although we would have to make sure that no maintenance tasks are running, which we don't have a way to do it as user may trigger those. Adding new command and documentation about how to manually schedule a Job doing an upgrade could work though.

New clusters have it enabled by default from 6.0, so nothing to do there.

piodul commented:
v1.1: rebased on master

piodul commented on docs/architecture/raft.rst:
```diff
@@ -54,6 +54,10 @@ When all the nodes in the cluster are upgraded to ScyllaDB Open Source 5.5 or Sc
 ========================================================================
 
+.. note::
+    After the procedure described in this section finishes, if your cluster is running ScyllaDB Open Source 6.0 then you must perform manual action in order to enable consistent topology changes.
```
> I think this is implicitly assumed by the fact that we version documentation

Right. However, I do see that we explicitly mention versions in some docs, e.g.:

```rst
When all the nodes in the cluster are upgraded to ScyllaDB Open Source 5.5 or ScyllaDB Enterprise 2024.2, the cluster will start the **Raft upgrade procedure**.
```

therefore I thought that this would be OK, but in the end I'm not sure. @annastuchlik , what's the best practice here?

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/17624)

**Changed Files:**
- [architecture/raft](http://previews.docs.scylladb.com/scylladb/17624/architecture/raft)
- [troubleshooting/handling-node-failures](http://previews.docs.scylladb.com/scylladb/17624/troubleshooting/handling-node-failures)
- [upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology](http://previews.docs.scylladb.com/scylladb/17624/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

piodul commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+* :doc:`Cluster management procedures </operating-scylla/procedures/cluster-management/index>` (adding, replacing, removing, decomissioning nodes etc.).
+* Runing :doc:`nodetool repair </operating-scylla/nodetool-commands/repair>`.
+* Runing :doc:`nodetool checkAndRepairCdcStreams </operating-scylla/nodetool-commands/checkandrepaircdcstreams>`.
+
```
I got conditioned to ignore it because it always shows some issues outside of my code...

piodul replied:
Actually, I didn't notice because I didn't look at the commit view / files changed view (I didn't really have a reason to do so). IMO scylla-promoter should indicate in its report that there are some new types that need to be looked at.

piodul commented:
v2:

- Removed the "if ScyllaDB 6.0" conditionals pointed out in review
- Rephrased some sentences to use simpler grammar
- Fixed spelling which was pointed out by the automatic check
- Adjusted examples according to suggestions
- Added information about checks that admin should perform before resorting to recovery in case upgrade gets stuck
- Mentioned the procedure for enabling consistent topology changes in the "You can follow the manual recovery procedure when" section of the document about handling node failures

piodul commented on docs/upgrade/upgrade-opensource/upgrade-guide-from-5.4-to-6.0/enable-consistent-topology.rst:
```diff
@@ -3,4 +3,102 @@ Enable Consistent Topology Updates
+Prerequisites
+=============
+
+* Make sure that all nodes in the cluster are upgraded to ScyllaDB Open Source 6.0.
```
How should I proceed with the docs for enterprise upgrade? Send a PR to the enterprise repo with "ScyllaDB Open Source 6.0" replaced with "ScyllaDB Enterprise 2024.2"?

kbr-scylla replied:
@annastuchlik In that case, would it be possible to turn this whole article into some kind of template living outside the `upgrade-guide-from-5.4-to-6.0` directory, and then this file would simply use that template and perform a bunch of variable substitutions?

(I don't know whether Sphinx supports this)
